### ISSUE TITLE
Information leakage? Decoder is taking in the full output sequence during evaluation?

### ISSUE THREAD

**Issue by GavinGuan95** [aimagelab/meshed-memory-transformer#17]

**Comment by baraldilorenzo** [comment]
Hi, could you please clarify what you mean, exactly, possibly with a reference to the code?

Best,
Lorenzo.

**Comment by GavinGuan95** [comment]
When calculating evaluate_loss in train.py, the model has access to the ground truth caption. In decoder.py MeshedDecoder forward, the variable "input" in "def forward(self, input, encoder_output, mask_encoder):" receives the ground truth caption regardless of whether it is training or validation. I might be missing something here.
Thank you for your time.

def evaluate_loss(model, dataloader, loss_fn, text_field):
    # Validation loss
    model.eval()
    running_loss = .0
    with tqdm(desc='Epoch %d - validation' % e, unit='it', total=len(dataloader)) as pbar:
        with torch.no_grad():
            for it, (detections, captions) in enumerate(dataloader):
                detections, captions = detections.to(device), captions.to(device)
                out = model(detections, captions)

**Comment by baraldilorenzo** [comment]
Hi,

of course having ground-truth data is mandatory for calculating a loss in a supervised setting, so the fact that the GT caption is an input to `evaluate_loss` is perfectly normal. Regarding the decoder, it only receives the GT caption during training, not during evaluation and testing. Take a look at what the `step` function [does](https://github.com/aimagelab/meshed-memory-transformer/blob/e0fe3fae68091970407e82e5b907cbc423f25df2/models/transformer/transformer.py#L36) when calling the decoder during the evaluation stage.

Best,
Lorenzo.
